pkg/prove/proof.go

```diff
@@ -22,99 +22,52 @@ import (
-	var rawShares [][]byte          //nolint:prealloc // rarely will this contain more than a single share
-	var rowRoots []tmbytes.HexBytes //nolint:prealloc // rarely will this contain more than a single root
-	for i, row := range rowShares {
-		// create an nmt to use to generate a proof
-		tree := wrapper.NewErasuredNamespacedMerkleTree(data.SquareSize, uint(i))
-		for _, share := range row {
-			tree.Push(
-				share,
-			)
-		}
-
-		startLeafPos := startLeaf
-		endLeafPos := endLeaf
-
-		// if this is not the first row, then start with the first leaf
-		if i > 0 {
-			startLeafPos = 0
-		}
-		// if this is not the last row, then select for the rest of the row
-		if i != (len(rowShares) - 1) {
-			endLeafPos = data.SquareSize - 1
-		}
-
-		rawShares = append(rawShares, shares.ToBytes(row[startLeafPos:endLeafPos+1])...)
-		proof, err := tree.Tree().ProveRange(int(startLeafPos), int(endLeafPos+1))
-		if err != nil {
-			return types.TxProof{}, err
-		}
-
-		proofs = append(proofs, &tmproto.NMTProof{
-			Start:    int32(proof.Start()),
-			End:      int32(proof.End()),
-			Nodes:    proof.Nodes(),
-			LeafHash: proof.LeafHash(),
-		})
-
-		// we don't store the data availability header anywhere, so we
-		// regenerate the roots to each row
-		rowRoots = append(rowRoots, tree.Root())
+	namespace := getTxNamespace(data.Txs[txIndex])
+	proof, err := NewShareInclusionProof(rawShares, data.SquareSize, namespace, startShare, endShare)
+	if err != nil {
+		return types.TxProof{}, err
 	}
 
 	return types.TxProof{
-		RowRoots: rowRoots,
-		Data:     rawShares,
-		Proofs:   proofs,
+		RowRoots: proof.RowsProof.RowsRoots,
+		Data:     proof.Data,
+		Proofs:   proof.SharesProofs,
 	}, nil
 }
 
+func getTxNamespace(tx types.Tx) (ns namespace.ID) {
+	_, isIndexWrapper := types.UnmarshalIndexWrapper(tx)
+	if isIndexWrapper {
+		return appconsts.PayForBlobNamespaceID
+	}
+	return appconsts.TxNamespaceID
+}
+
 // TxSharePosition returns the start and end positions for the shares that
 // include a given txIndex. Returns an error if index is greater than the length
 // of txs.
-func TxSharePosition(txs types.Txs, txIndex uint64) (startSharePos, endSharePos uint64, err error) {
-	if txIndex >= uint64(len(txs)) {
-		return startSharePos, endSharePos, errors.New("transaction index is greater than the number of txs")
+func TxSharePosition(data types.Data, txIndex uint64) (startShare uint64, endShare uint64, err error) {
+	if int(txIndex) >= len(data.Txs) {
+		return 0, 0, errors.New("transaction index is greater than the number of txs")
 	}
 
-	prevTxTotalLen := 0
-	for i := uint64(0); i < txIndex; i++ {
-		txLen := len(txs[i])
-		prevTxTotalLen += (shares.DelimLen(uint64(txLen)) + txLen)
-	}
-
-	currentTxLen := len(txs[txIndex])
-	currentTxTotalLen := shares.DelimLen(uint64(currentTxLen)) + currentTxLen
-	endOfCurrentTxLen := prevTxTotalLen + currentTxTotalLen
+	_, _, shareRanges := shares.SplitTxs(data.Txs)
```

Comment:
Is there ever the case where we want the shares and the share ranges i.e. what is the case for having both of these values returned by the same method and not having different methods for each? (other than the extra lines of code)

Reply:
Good point. I don't see a scenario where they both need to be returned at the same time. I'll consider extracting to a stand-alone function

Reply:
I took a stab at extracting a separate function like this:

```go
func TxShareRanges(txs coretypes.Txs) (shareRanges map[coretypes.TxKey]ShareRange) {
	txWriter := NewCompactShareSplitter(appconsts.TxNamespaceID, appconsts.ShareVersionZero)
	pfbTxWriter := NewCompactShareSplitter(appconsts.PayForBlobNamespaceID, appconsts.ShareVersionZero)

	for _, tx := range txs {
		if _, isIndexWrapper := coretypes.UnmarshalIndexWrapper(tx); isIndexWrapper {
			pfbTxWriter.WriteTx(tx)
		} else {
			txWriter.WriteTx(tx)
		}
	}

	txShares, txMap := txWriter.Export(0)
	_, pfbMap := pfbTxWriter.Export(len(txShares))

	return mergeMaps(txMap, pfbMap)
}
```

It shares a lot of the same implementation with SplitTxs but makes it so that the SplitTxs function signature can be simpler:

```diff
-func SplitTxs(txs coretypes.Txs) (txShares []Share, pfbShares []Share, shareRanges map[coretypes.TxKey]ShareRange) {
+func SplitTxs(txs coretypes.Txs) (txShares []Share, pfbShares []Share, shareRanges) {
```

the downside to this approach is that we have duplicate logic in the bodies of these functions and I'm nervous that the implementation of `SplitTxs` evolves without a corresponding update to `TxShareRanges` - does that concern make sense?